src/ray/gcs/gcs_server/gcs_actor_scheduler.cc

```diff
@@ -0,0 +1,334 @@
+    // Reset the client so that we can reconnect when retry.
+    ResetClient();
+    // The purpose of using the weak_ptr here is to make the worker's life cycle easier to
+    // manage.
+    // Capturing the shared_ptr in lambda is another choice, but it is a little more
+    // complexity because it need add a flag to the worker, if the worker is dead we
+    // should mark the flag to dead, and in the callback of io event we should do like
+    // this:
+    //    if (!shared_worker->is_dead) { do_something(); }
+    std::weak_ptr<GcsLeasedWorker> weak_this(shared_from_this());
+    execute_after(io_context_,
+                  [weak_this, actor, on_done] {
+                    if (auto worker = weak_this.lock()) {
+                      worker->CreateActor(actor, on_done);
+                    }
+                  },
+                  RayConfig::instance().gcs_create_actor_retry_interval_ms());
+  };
+
+  std::weak_ptr<GcsLeasedWorker> weak_this(shared_from_this());
+  auto client = GetOrCreateClient();
+  auto status = client->PushNormalTask(
+      std::move(request), [weak_this, retry_creating_actor, on_done](
+                              Status status, const rpc::PushTaskReply &reply) {
+        RAY_UNUSED(reply);
+        if (auto worker = weak_this.lock()) {
+          if (!status.ok()) {
+            retry_creating_actor();
+            return;
+          }
+          on_done();
+        }
+      });
+  if (!status.ok()) {
+    retry_creating_actor();
+  }
+}
+
+GcsActorScheduler::GcsActorScheduler(boost::asio::io_context &io_context,
+                                     gcs::GcsNodeManager &gcs_node_manager)
+    : io_context_(io_context),
+      client_call_manager_(io_context_),
+      gcs_node_manager_(gcs_node_manager) {}
+
+void GcsActorScheduler::Schedule(std::shared_ptr<GcsActor> actor) {
+  auto node_id = actor->GetNodeID();
+  if (!node_id.IsNil()) {
+    if (auto node = gcs_node_manager_.GetNode(node_id)) {
+      // If the actor is already tied to a node and the node is available, then record
+      // the relationship of the node and actor and then lease worker directly from the
+      // node.
+      AddActorInPhaseOfLeasing(actor);
+      LeaseWorkerFromNode(actor, node);
+      return;
+    }
+
+    // The actor is already tied to a node which is unavailable now, so we should reset
+    // the address.
+    actor->UpdateAddress(rpc::Address());
+  }
+
+  // Select a node to lease worker for the actor.
+  auto node = SelectNodeRandomly();
+  if (node == nullptr) {
+    // There are no available nodes to schedule the actor, so just trigger the failed
+    // handler.
+    RAY_CHECK(schedule_failed_handler_);
+    schedule_failed_handler_(std::move(actor));
+    return;
+  }
+
+  // Update the address of the actor as it is tied to a new node.
+  auto address = actor->GetAddress();
+  address.set_raylet_id(node->node_id());
+  actor->UpdateAddress(address);
+  actor->AsyncFlush([this, actor] {
```

Comment:
> Why do we need to flush every time a new node is picked? Isn't it fine to just pick a different node if the GCS service restarts?

We must ensure that every change of the node needs to be flushed. Consider such a scenario. If we choose Node1 to create Actor1 at the beginning, and the resource of Node1 is not enough and spill back a Node2. Then the GcsActorScheduler will try to create Actor1 at Node2. Suddenly then GCS crash and restart, it will load the data of the previous flushed to the storage. If the change is not written to the storage before creating Actor1 on Node2, the GcsActorScheduler will create Actor1 on Node1, which will cause the Actor1 be created on both Node1 and Node2.